pkg/function/service.go

```diff
@@ -221,28 +218,40 @@ func (fs *Service) ReduceFn(stream functionpb.UserDefinedFunction_ReduceFnServer
 			ch = make(chan Datum)
 			chanMap[d.Key] = ch
 
-			wg.Add(1)
-			go func(key string, ch chan Datum) {
-				defer wg.Done()
-				messages := fs.Reducer.HandleDo(ctx, key, ch, md)
-				mu.Lock()
-				defer mu.Unlock()
-				for _, msg := range messages {
-					datumList = append(datumList, &functionpb.Datum{
-						Key:   msg.Key,
-						Value: msg.Value,
-					})
-				}
+			func(key string, ch chan Datum) {
+				g.Go(func() error {
+					messages := fs.Reducer.HandleDo(ctx, key, ch, md)
+					datumList := buildDatumList(messages)
+
+					// stream.Send() is not thread safe.
+					mu.Lock()
+					defer mu.Unlock()
+					err := stream.Send(datumList)
```

Comment:
This is assuming the returned datum list is not over the size limit?

Reply:
yeah, if it's over the size limit it will return an error